File: examples/bun/src/layout-flex.ts
/**
 * Flexbox: every direction, both wraps, and each way of distributing a line.
 *
 * Rows of coloured blocks rather than a picture, because the question each row
 * answers is *where things sit*. A block that fails to move is visible against
 * its neighbours; a prettier scene would hide it.
 */

import { Box, Root, Row, type Children, type SceneNode } from 'meo-canvas'

import { FORMATS, draw } from './write.js'

const RED = '#dc2828'
const BLUE = '#2850dc'
const GREEN = '#288c3c'

/** One coloured block of a fixed size. */
const block = (colour: string, width: number, height = 18): SceneNode => Box({ width, height, backgroundColor: colour })

/** A strip with a background, so an empty row is still visible. */
const strip = (children: Children, rest: Record<string, unknown> = {}): SceneNode =>
  Row({ width: 180, height: 26, padding: 4, gap: 4, backgroundColor: '#eeeef2', children, ...rest })

const canvas = await Root({
  width: 400,
  height: 372,
  backgroundColor: '#ffffff',
  padding: 8,
  flexDirection: 'column',
  gap: 6,
  children: [
    // Direction: the same three blocks, read two ways.
    strip([block(RED, 30), block(BLUE, 30), block(GREEN, 30)]),
    strip([block(RED, 30), block(BLUE, 30), block(GREEN, 30)], { flexDirection: 'row-reverse' }),
    // Justify: where the free space goes along the main axis.
    strip([block(RED, 24), block(BLUE, 24)], { justifyContent: 'space-between' }),
    strip([block(RED, 24), block(BLUE, 24)], { justifyContent: 'center' }),
    strip([block(RED, 24), block(BLUE, 24)], { justifyContent: 'space-evenly' }),
    // Align: where a shorter item sits across the line.
    strip([block(RED, 24, 8), block(BLUE, 24, 18)], { height: 30, alignItems: 'flex-end' }),
    strip([block(RED, 24, 8), block(BLUE, 24, 18)], { height: 30, alignItems: 'center' }),
    // Grow, shrink and basis: the three ways a length is negotiated.
    strip([
      Box({ width: 20, height: 18, backgroundColor: RED, flexGrow: 1 }),
      block(BLUE, 20),
      Box({ width: 20, height: 18, backgroundColor: GREEN, flexGrow: 2 }),
    ]),
    strip([Box({ width: 200, height: 18, backgroundColor: RED, flexShrink: 1 }), Box({ width: 200, height: 18, backgroundColor: BLUE, flexShrink: 3 })]),
    // Wrap: a line that cannot hold its children.
    strip([block(RED, 60), block(BLUE, 60), block(GREEN, 60)], { height: 52, flexWrap: 'wrap' }),
    // Aspect ratio: a height derived from a width.
    Row({
      width: 180,
      height: 26,
      backgroundColor: '#eeeef2',
      flexDirection: 'column',
      children: Box({ width: 48, aspectRatio: 3, backgroundColor: GREEN }),
    }),
  ],
})

await draw('layout-flex', canvas, FORMATS)

File: examples/rust/src/bin/layout-flex.rs
//! Flexbox: every direction, both wraps, and each way of distributing a line.
//!
//! Rows of coloured blocks rather than a picture, because the question each row
//! answers is *where things sit*. A block that fails to move is visible against
//! its neighbours; a prettier scene would hide it.

use meo_canvas::{
    Align, Box, Column, FlexDirection, FlexWrap, Justify, Root, Row, Styled,
    hex_rgb, px,
};
use meo_canvas_examples::{FORMATS, draw};

/// One coloured block of a fixed size.
fn block(colour: u32, width: f32) -> meo_canvas::Element {
    Box::new()
        .size(px(width), px(18.0))
        .background_color(hex_rgb(colour))
}

/// A labelled strip with a background, so an empty row is still visible.
fn strip(children: impl meo_canvas::IntoElements) -> meo_canvas::Element {
    Row::new()
        .size(px(180.0), px(26.0))
        .padding(px(4.0))
        .gap(px(4.0))
        .background_color(hex_rgb(0xee_ee_f2))
        .children(children)
}

fn main() -> Result<(), std::boxed::Box<dyn std::error::Error>> {
    let red = 0xdc_28_28;
    let blue = 0x28_50_dc;
    let green = 0x28_8c_3c;

    let root = Root::new(400.0)
        .height(372.0)
        .background_color(hex_rgb(0xff_ff_ff))
        .padding(px(8.0))
        .flex_direction(FlexDirection::Column)
        .gap(px(6.0))
        .children([
            // Direction: the same three blocks, read four ways.
            strip(vec![
                block(red, 30.0),
                block(blue, 30.0),
                block(green, 30.0),
            ]),
            strip(vec![
                block(red, 30.0),
                block(blue, 30.0),
                block(green, 30.0),
            ])
            .flex_direction(FlexDirection::RowReverse),
            // Justify: where the free space goes along the main axis.
            strip(vec![block(red, 24.0), block(blue, 24.0)])
                .justify_content(Justify::SpaceBetween),
            strip(vec![block(red, 24.0), block(blue, 24.0)])
                .justify_content(Justify::Center),
            strip(vec![block(red, 24.0), block(blue, 24.0)])
                .justify_content(Justify::SpaceEvenly),
            // Align: where a shorter item sits across the line.
            strip(vec![
                block(red, 24.0).height(px(8.0)),
                block(blue, 24.0).height(px(18.0)),
            ])
            .height(px(30.0))
            .align_items(Align::FlexEnd),
            strip(vec![
                block(red, 24.0).height(px(8.0)),
                block(blue, 24.0).height(px(18.0)),
            ])
            .height(px(30.0))
            .align_items(Align::Center),
            // Grow, shrink and basis: the three ways a length is negotiated.
            strip(vec![
                block(red, 20.0).flex_grow(1.0),
                block(blue, 20.0),
                block(green, 20.0).flex_grow(2.0),
            ]),
            strip(vec![
                block(red, 200.0).flex_shrink(1.0),
                block(blue, 200.0).flex_shrink(3.0),
            ]),
            // Wrap: a line that cannot hold its children.
            strip(vec![
                block(red, 60.0),
                block(blue, 60.0),
                block(green, 60.0),
            ])
            .height(px(52.0))
            .flex_wrap(FlexWrap::Wrap),
            // Aspect ratio: a height derived from a width.
            Column::new()
                .size(px(180.0), px(26.0))
                .background_color(hex_rgb(0xee_ee_f2))
                .children(
                    Box::new()
                        .width(px(48.0))
                        .aspect_ratio(3.0)
                        .background_color(hex_rgb(green)),
                ),
        ]);

    draw("layout-flex", root, FORMATS)
}
